src/AdoPat/PatManager.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AdoPat
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.Services.DelegatedAuthorization;
+
+    /// <summary>
+    /// Manage the interaction between the PAT cache and Azure DevOps.
+    /// </summary>
+    public class PatManager
+    {
+        private const int ValidToExtensionDays = 7;
+        private const int NearingExpirationDays = 2;
+
+        private IPatCache cache;
+        private IPatClient client;
+        private Func<DateTime> now;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PatManager"/> class.
+        /// </summary>
+        /// <param name="cache">Any class that implements the <see cref="IPatCache"/> interface.</param>
+        /// <param name="client">Any class that implements the <see cref="IPatClient"/> interface.</param>
+        /// <param name="now">A function for computing the current moment. Defaults to null, which
+        /// uses <see cref="DateTime.UtcNow"/>. Overriding this should only be necessary in testing.</param>
+        public PatManager(IPatCache cache, IPatClient client, Func<DateTime> now = null)
+        {
+            this.cache = cache;
+            this.client = client;
+            this.now = now ?? new Func<DateTime>(() => DateTime.UtcNow);
+        }
+
+        /// <summary>
+        /// Fetches a PAT matching the given options from the cache, creating or regenerating a new PAT as necessary.
+        /// </summary>
+        /// <param name="options">Options used to match the PAT in the cache or when creating/regenerating a new one.</param>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>An Azure DevOps Personal Access Token.</returns>
+        public async Task<PatToken> GetPatAsync(
+            PatOptions options,
+            CancellationToken cancellationToken = default)
+        {
+            bool writeBack = false;
+            var pat = this.cache.GetPat(options.ToString());

```

Comment:
I have a slightly odd feeling about using ToString as the role of the `key`. ToString is designed for the string representation, typically a readable one. Here, the string representation you've chosen is motivated specifically to be the Key. In the future I could see someone easily changing the ToString to be more human readable, and it's not encoded or obvious that that output is actually what is used for the cache keys. 

I think I'd prefer the ToString to focus on it's role, if we even want one - for displaying on stdout the Pat, and specically add a `CacheKey` method to return the string to be used for CacheKeying.

Reply:
Fair enough. More than happy to just rename this method.